runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -110,6 +111,286 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
+   deps->addPostCondition(x0Reg, TR::RealRegister::x0);
+   deps->addPostCondition(metaReg, TR::RealRegister::NoReg);
+
+   tempMR = new (cg->trHeapMemory()) TR::MemoryReference(node, 4, cg);
+
+   node->setRegister(objReg);
+
+   if (tempMR->getUnresolvedSnippet() != NULL)
+      {
+      generateTrg1MemInstruction(cg, TR::InstOpCode::addx, node, locationReg, tempMR);
+      }
+   else
+      {
+      if (tempMR->useIndexedForm())
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::addx, node, locationReg, tempMR->getBaseRegister(), tempMR->getIndexRegister());
+      else
+         generateTrg1MemInstruction(cg, TR::InstOpCode::addimmx, node, locationReg, tempMR);
+      }
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmw, node, objReg, new (cg->trHeapMemory()) TR::MemoryReference(locationReg, 0, cg));
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, startLabel);
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmw, node, evacuateReg,
+         new (cg->trHeapMemory()) TR::MemoryReference(metaReg, cg->comp()->fej9()->thisThreadGetEvacuateBaseAddressOffset(), cg));
+   generateCompareInstruction(cg, node, objReg, evacuateReg, false); // false for 32-bit comparison
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, endLabel, TR::CC_LT);
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmw, node, evacuateReg,
+         new (cg->trHeapMemory()) TR::MemoryReference(metaReg, cg->comp()->fej9()->thisThreadGetEvacuateTopAddressOffset(), cg));
+   generateCompareInstruction(cg, node, objReg, evacuateReg, false); // false for 32-bit comparison
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, endLabel, TR::CC_GT);
+
+   // TR_softwareReadBarrier helper expects the vmThread in x0.
+   generateMovInstruction(cg, node, x0Reg, metaReg);
+
+   TR::SymbolReference *helperSym = comp->getSymRefTab()->findOrCreateRuntimeHelper(TR_softwareReadBarrier, false, false, false);
+   generateImmSymInstruction(cg, TR::InstOpCode::bl, node, (uintptr_t)helperSym->getMethodAddress(), deps, helperSym, NULL);
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmw, node, objReg, new (cg->trHeapMemory()) TR::MemoryReference(locationReg, 0, cg));
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, endLabel, deps);
+
+   bool needSync = (node->getSymbolReference()->getSymbol()->isSyncVolatile() && cg->comp()->target().isSMP());
+
+   if (needSync)
+      {
+      generateSynchronizationInstruction(cg, TR::InstOpCode::dsb, node, 0xF); // dsb SY
+      }
+
+   cg->insertPrefetchIfNecessary(node, objReg);
+
+   tempMR->decNodeReferenceCounts(cg);
+
+   cg->stopUsingRegister(evacuateReg);
+   cg->stopUsingRegister(locationReg);
+   cg->stopUsingRegister(x0Reg);
+
+   cg->machine()->setLinkRegisterKilled(true);
+
+   return objReg;
+#endif // OMR_GC_CONCURRENT_SCAVENGER
+   }
+
+static TR::Register *
+aGenerateSoftwareReadBarrier(TR::Node *node, TR::CodeGenerator *cg)
```

Comment:
The DNA of this method is about 80-90% identical to the iGenerateSoftwareReadBarrier method.  Is there any way you could combine them to save some code space?

Reply:
Most of my comments in iGenerateSoftwareReadBarrier apply to this function as well so I'm not going to repeat them.

Reply:
Merged the two functions into `generateSoftwareReadBarrier()`, which takes `bool isArdbari` as its third argument.